File: src/main/java/net/dashflight/data/helpers/Memoizer.java
package net.dashflight.data.helpers;

import java.util.concurrent.Callable;
import java.util.concurrent.CancellationException;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import java.util.concurrent.FutureTask;


public class Memoizer<A, V> implements Computable<A, V> {

    private final ConcurrentMap<A, Future<V>> cache = new ConcurrentHashMap<>();

    private final Computable<A, V> c;

    public Memoizer(Computable<A, V> c) {
        this.c = c;
    }

    public V compute(final A arg) throws InterruptedException {
        while (true) {

            Future<V> f = cache.get(arg);
            // computation not started
            if (f == null) {
                System.out.println("Computation not started");
                Callable<V> eval = () -> c.compute(arg);

                FutureTask<V> ft = new FutureTask<>(eval);
                f = cache.putIfAbsent(arg, ft);
                // start computation if it's not started in the meantime
                if (f == null) {
                    System.out.println("Computation beginning...");

                    f = ft;
                    ft.run();
                }
            } else {
                System.out.println("Computation already started!");
            }

            // get result if ready, otherwise block and wait
            try {
                System.out.println("Waiting for result...");
                return f.get();
            } catch (CancellationException e) {
                cache.remove(arg, f);
            } catch (ExecutionException e) {
                e.printStackTrace();
                throw new RuntimeException(e.getCause());
            }
        }
    }
}

File: src/main/java/net/dashflight/data/caching/Memoizer.kt
package net.dashflight.data.caching

import java.util.concurrent.*

/**
 * Prevents many concurrent calls from independently fetching the same data.
 *
 * If computation is not running, begin and wrap result in a Future for other callers.
 * If computation is running, caller will wait for computation to finish and collect result.
 */
class Memoizer<K, V>(private val computeFunction: Computable<K, V>) : Computable<K, V?> {
    private val cache: ConcurrentMap<K, Future<V?>?> = ConcurrentHashMap()

    @Throws(InterruptedException::class)
    override fun compute(key: K): V? {
        while (true) {
            var startedComputation = false
            var future = cache[key]

            // Computation not started
            if (future == null) {
                val eval = Callable { computeFunction.compute(key) }
                val futureTask = FutureTask(eval)
                future = cache.putIfAbsent(key, futureTask)

                // Start computation if it's not started in the meantime
                if (future == null) {
                    future = futureTask
                    futureTask.run()
                    startedComputation = true
                }
            }

            // Get result if ready, otherwise block and wait
            try {
                val result = future.get()

                // Remove future from map to prevent caching value forever
                if (startedComputation) {
                    cache.remove(key)
                }
                return result
            } catch (e: CancellationException) {
                cache.remove(key, future)
            } catch (e: ExecutionException) {
                e.printStackTrace()
                throw RuntimeException(e.cause)
            }
        }
    }

}